axelor-production/src/main/java/com/axelor/apps/production/service/bomimport/BillOfMaterialImportServiceImpl.java

```diff
@@ -0,0 +1,195 @@
+      BillOfMaterialImporter billOfMaterialImporter) {
+    this.appBaseService = appBaseService;
+    this.importService = importService;
+    this.billOfMaterialLineService = billOfMaterialLineService;
+    this.importConfigurationRepository = importConfigurationRepository;
+    this.billOfMaterialImportRepository = billOfMaterialImportRepository;
+    this.billOfMaterialRepository = billOfMaterialRepository;
+    this.billOfMaterialImporter = billOfMaterialImporter;
+  }
+
+  @Override
+  public ImportHistory processImport(BillOfMaterialImport billOfMaterialImport)
+      throws AxelorException, IOException {
+    billOfMaterialImporter.addBillOfMaterialImport(billOfMaterialImport);
+    return billOfMaterialImporter.init(createImportConfiguration(billOfMaterialImport)).run();
+  }
+
+  @Transactional
+  protected ImportConfiguration createImportConfiguration(
+      BillOfMaterialImport billOfMaterialImport) {
+    ImportConfiguration importConfiguration = new ImportConfiguration();
+
+    importConfiguration.setUser(AuthUtils.getUser());
+    importConfiguration.setName(billOfMaterialImport.getName());
+    importConfiguration.setBindMetaFile(billOfMaterialImport.getImportSource().getBindingFile());
+    importConfiguration.setDataMetaFile(billOfMaterialImport.getImportMetaFile());
+    importConfiguration.setStartDateTime(appBaseService.getTodayDateTime().toLocalDateTime());
+
+    return importConfigurationRepository.save(importConfiguration);
+  }
+
+  @Override
+  @Transactional
+  public BillOfMaterialImport setStatusToImported(BillOfMaterialImport billOfMaterialImport) {
+    billOfMaterialImport = billOfMaterialImportRepository.find(billOfMaterialImport.getId());
+    billOfMaterialImport.setStatusSelect(BillOfMaterialImportRepository.STATUS_IMPORTED);
+    return billOfMaterialImportRepository.save(billOfMaterialImport);
+  }
+
+  @Override
+  public BillOfMaterial createBoMFromImport(BillOfMaterialImport billOfMaterialImport)
+      throws AxelorException {
+    if (billOfMaterialImport != null
+        && billOfMaterialImport.getMainBillOfMaterialGenerated() != null) {
+      return null;
+    }
+    Optional<BillOfMaterialImportLine> billOfMaterialImportLineOp = Optional.empty();
+    if (billOfMaterialImport != null
+        && billOfMaterialImport.getBillOfMaterialImportLineList() != null) {
+      billOfMaterialImportLineOp =
+          billOfMaterialImport.getBillOfMaterialImportLineList().stream()
+              .filter(line -> line.getBomLevel() == 0)
+              .findFirst();
+    }
+    if (billOfMaterialImportLineOp.isEmpty()) {
+      throw new AxelorException(
+          billOfMaterialImport,
+          TraceBackRepository.CATEGORY_MISSING_FIELD,
+          ProductionExceptionMessage.BOM_IMPORT_NO_MAIN_BILL_OF_MATERIAL_GENERATED);
+    }
+    BillOfMaterialImportLine billOfMaterialImportLine = billOfMaterialImportLineOp.get();
+
+    return generateBillOfMaterialFromImportLine(billOfMaterialImportLine).orElse(null);
```

Comment:
can be done in one line
something like billOfMaterialImportLineOp.map(boImportLine -> generateBillOfMaterialFromImportLine(boImportLine)).orElse()